packages/core/contracts/financial-templates/optimistic-rewarder/OptimisticRewarder.sol

```diff
@@ -0,0 +1,429 @@
+                ancillaryData,
+                request,
+                msg.sender,
+                address(this)
+            );
+        }
+
+        delete redemptions[redemptionId];
+    }
+
+    /**
+     * @notice Redeem a redemption request that has passed liveness.
+     * @dev returns the bond that was paid with the initial proposal.
+     * @param tokenId the tokenId the redemption is for.
+     * @param cumulativeRedemptions the cumulative redemptions that were provided in the original request.
+     */
+    function redeem(uint256 tokenId, RedemptionAmount[] memory cumulativeRedemptions) public nonReentrant {
+        bytes32 redemptionId = _redemptionId(tokenId, cumulativeRedemptions);
+
+        // Can only be redeemed by owner.
+        require(msg.sender == ownerOf(tokenId), "must be called by token owner");
+
+        // Check that the redemption is initialized and that it passed liveness.
+        require(
+            redemptions[redemptionId].expiryTime != 0 && getCurrentTime() >= redemptions[redemptionId].expiryTime,
+            "unexpired or nonexistent"
+        );
+
+        for (uint256 i = 0; i < cumulativeRedemptions.length; i++) {
+            IERC20 token = cumulativeRedemptions[i].token;
+            uint256 currentRedemptionTotal = redeemedAmounts[tokenId][token];
+            uint256 proposedRedemptionTotal = cumulativeRedemptions[i].amount;
+
+            // Only pay if the cumulative amount specified in the request is larger than the amount paid out already.
+            if (proposedRedemptionTotal > currentRedemptionTotal) {
+                uint256 amountToPay = proposedRedemptionTotal - currentRedemptionTotal;
+                redeemedAmounts[tokenId][token] = proposedRedemptionTotal;
+                token.safeTransfer(msg.sender, amountToPay);
+            }
+        }
+
+        // Return the bond to the owner.
+        bondToken.safeTransfer(msg.sender, redemptions[redemptionId].expiryTime + bond);
+
+        delete redemptions[redemptionId];
+
+        emit Redeemed(tokenId, redemptionId, redemptions[redemptionId].expiryTime);
+    }
+
+    /**
+     * @notice gets the current time. Can be overridden for testing.
+     */
+    function getCurrentTime() public view virtual returns (uint256) {
+        return block.timestamp;
+    }
+
+    /**
+     * @notice Abstract function that is called to mint the next ERC721 tokenId.
+     * @param recipient the recipient of the newly minted token.
+     */
+    function mintNextToken(address recipient) public virtual returns (uint256);
+
+    /**
+     * @notice Abstract function that is called to check the owner of the token.
+     * @dev this matches the ERC721 ownerOf interface.
+     * @param tokenId the tokenId to check the owner of.
+     */
+    function ownerOf(uint256 tokenId) public view virtual returns (address);
+
+    function _redemptionId(uint256 tokenId, RedemptionAmount[] memory amounts) internal pure returns (bytes32) {
```

Comment:
any downside to making this public? or make a nonReentrant view function that lets you call this? will just make things a bit easier to re-compute off-chain to have ea view function you can easily call for this.

Reply:
Good point. Will do

Reply:
Done.